torch_xla/csrc/tensor.h

```diff
@@ -53,31 +53,43 @@ using XLATensorPtr = c10::intrusive_ptr<XLATensor>;
 
 class XLATensor : public torch::lazy::LazyTensor {
  public:
+  struct ShardingSpec;
+  using ShardingSpecPtr = std::shared_ptr<ShardingSpec>;
+
   // This is the core XLA tensor data structure where all the tensor data is
   // held. The XLA tensor is nothing more than a shared pointer to a Data
   // object.
   struct Data : public torch::lazy::LazyTensor::Data {
     Data(torch::lazy::BackendDataPtr handle,
          const torch::lazy::BackendDevice& device,
-         c10::optional<at::ScalarType> logical_element_type)
+         c10::optional<at::ScalarType> logical_element_type,
+         ShardingSpecPtr sharding = nullptr)
         : torch::lazy::LazyTensor::Data(handle, device),
-          logical_element_type(logical_element_type) {}
+          logical_element_type(logical_element_type),
+          sharding(sharding) {}
     Data(torch::lazy::Value ir_value, const torch::lazy::BackendDevice& device,
-         c10::optional<at::ScalarType> logical_element_type)
+         c10::optional<at::ScalarType> logical_element_type,
+         ShardingSpecPtr sharding = nullptr)
         : torch::lazy::LazyTensor::Data(ir_value, device),
-          logical_element_type(logical_element_type) {}
-    Data(at::Tensor tensor_data, const torch::lazy::BackendDevice& device)
+          logical_element_type(logical_element_type),
+          sharding(sharding) {}
+    Data(at::Tensor tensor_data, const torch::lazy::BackendDevice& device,
+         ShardingSpecPtr sharding = nullptr)
         : torch::lazy::LazyTensor::Data(tensor_data, device),
-          logical_element_type(tensor_data.scalar_type()) {}
+          logical_element_type(tensor_data.scalar_type()),
+          sharding(sharding) {}
     Data(std::shared_ptr<View> view, const torch::lazy::BackendDevice& device,
-         c10::optional<at::ScalarType> logical_element_type)
+         c10::optional<at::ScalarType> logical_element_type,
+         ShardingSpecPtr sharding = nullptr)
         : torch::lazy::LazyTensor::Data(device),
           view(std::move(view)),
-          logical_element_type(logical_element_type) {}
+          logical_element_type(logical_element_type),
+          sharding(sharding) {}
 
     ~Data();
 
     std::shared_ptr<View> view;
+    ShardingSpecPtr sharding;
```

Comment:
can we add comment here to explain this sharding is the source of truth and the ir one is just a copy?

Reply:
Sounds good.